Issue #5304 HTTP2 HostHeader

gregw opened the pull request:
#5304 
 Updated HostHeaderCustomizer to actually add the Host header, either from values passed in the custructor or from the getServerName and getServerPort methods.

The HttpURI is no longer updated.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #5304 HTTP2 HostHeader

Updated HostHeaderCustomizer to actually add the Host header, either from values passed in the custructor or from the getServerName and getServerPort methods.

The HttpURI is no longer updated.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

joakime commented on jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java:
```diff
@@ -19,30 +19,30 @@
     private final String serverName;
     private final int serverPort;
 
+    public HostHeaderCustomizer()
```
Some javadoc here saying this usage will result in using the serverName/serverPort from the Request?

joakime commented on jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java:
```diff
@@ -19,30 +19,30 @@
 
+    public HostHeaderCustomizer()
+    {
+        this(null, 0);
```
This forwards to `this(String, int)` which requires that `serverName` is not null.

joakime reviewed (changes requested):
This probably needs test cases, as there is a flaw in the null serverName case in this code.

gregw committed:
Issue #5304 HTTP2 HostHeader

 + Found and fixed bug in HttpFields
 + Added port normalization support to HttpScheme
 + added test

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
blank line

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #5304 HTTP2 HostHeader

 + refixed bug in HttpFields

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #5304 HTTP2 HostHeader

 + still fixing HttpFields bug

Signed-off-by: Greg Wilkins <gregw@webtide.com>

joakime commented on jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java:
```diff
@@ -57,15 +60,26 @@ public HostHeaderCustomizer(String serverName)
+            String host = serverName == null ? request.getServerName() : serverName;
+            int port = HttpScheme.normalizePort(request.getScheme(), serverPort == 0 ? request.getServerPort() : serverPort);
+
+            if (serverName != null || serverPort > 0)
```
Don't you want to setHttpURI() if the `port` is normalized in a way that's different then `serverPort` too?

gregw replied:
My thinking here is that if the serverPort is set, then we will set the port, but will used the normalised version of it.
So if they set port 80, but the real port was 8888 we would set the port to 0 not 80

joakime commented on jetty-server/src/main/java/org/eclipse/jetty/server/HostHeaderCustomizer.java:
```diff
@@ -57,15 +60,26 @@ public HostHeaderCustomizer(String serverName)
+            HttpFields original = request.getHttpFields();
+            HttpFields.Mutable httpFields = HttpFields.build(original.size() + 1);
+            httpFields.add(new HostPortHttpField(host, port));
+            httpFields.add(request.getHttpFields());
```
This looks odd, why is this not `httpFields.addAll(request.getHttpFields());` ?

Also, doesn't this change the `request.getHttpFields()` to now be Mutable?

What if this was ...

``` java
HttpFields updated = HttpFields.build(
   request.getHttpFields(), 
   new HostPortHttpFields(host, port));
request.setHttpFields(updated);
```

Where there is a new method `.build(HttpFields original, HttpField ... fields)` ?

gregw replied:
https://tools.ietf.org/html/rfc7230#section-3.2.2 says Host header should be one of the first headers sent.  I can't imagine too much depending on this, but best to follow recommendations.

joakime replied:
This still changes future requests on `request.getHttpFields()` to be a `HttpFields.Mutable`, right?
I don't think that's desired.

gregw replied:
https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java#L250-L253

joakime reviewed (changes requested):
Some minor concerns.

gregw committed:
Issue #5304 HTTP2 Host Header

updates from review

gregw commented:
@joakime nudge

joakime reviewed (changes requested):
accidentally Mutable?

gregw commented:
@joakim, it's not accidentally mutable. The style we mostly use is that the setter decides if it wants the mutable or immutable reference.  So in this case request will call asImutable on the passed HttpField.

The idea is that the callee decides and the caller can never make the request fields mutable

joakime reviewed (approved):
LGTM
